File: go/verify/references.go
// SPDX-License-Identifier: BSD-3-Clause

package verify

import "fmt"

// referenceFindings checks draft-04 §5.5.5 without resolving cited artifacts.
// CPB owns digest representation and comparison context, so a foreign reference
// must not be constrained to AAC's SHA-256 hex encoding. Log coordinates are
// recorded claims only; Class 1 never authenticates their inclusion proofs.
func referenceFindings(capsule map[string]interface{}, known map[string]map[string]bool) []Finding {
	if capsule["format_version"] != "4" {
		return nil // references is a draft-04 addition; preserve vintage extension handling.
	}
	raw, present := capsule["references"]
	if !present {
		return nil
	}
	var findings []Finding
	add := func(code, detail string, check int, severity string) {
		findings = append(findings, Finding{Code: code, Detail: detail, Check: mkCheck(check), Severity: severity})
	}
	refs, ok := raw.([]interface{})
	if !ok {
		add("references_malformed", "references MUST be an array (§5.5.5)", 1, "error")
		return findings
	}
	chain := asMap(capsule["chain"])
	parent, _ := chain["parent_capsule_id"].(string)
	for i, rawRef := range refs {
		path := fmt.Sprintf("references[%d]", i)
		ref := asMap(rawRef)
		if ref == nil {
			add("reference_malformed", path+" MUST be an object (§5.5.5)", 1, "error")
			continue
		}
		for _, field := range []string{"type", "digest_alg", "digest"} {
			if value, ok := ref[field].(string); !ok || value == "" {
				add("reference_malformed", path+"."+field+" MUST be a non-empty string (§5.5.5)", 1, "error")
			}
		}
		// Compare only a known AAC identity context, never equal-looking digests
		// belonging to a different artifact type or hash algorithm (CPB §7).
		if ref["type"] == "agent-action-capsule" && ref["digest_alg"] == "SHA-256" && !isHex64(ref["digest"]) {
			add("reference_malformed", path+".digest MUST be an AAC Capsule ID for agent-action-capsule/SHA-256 (§5.5.5)", 1, "error")
		}
		if ref["type"] == "agent-action-capsule" && ref["digest_alg"] == "SHA-256" && parent != "" && ref["digest"] == parent {
			add("reference_duplicates_chain_parent", path+" duplicates chain.parent_capsule_id (§5.5.5)", 6, "error")
		}
		if rawPurpose, present := ref["citation_purpose"]; present {
			purpose, ok := rawPurpose.(string)
			if !ok || purpose == "" {
				add("reference_malformed", path+".citation_purpose MUST be a non-empty string (§5.5.5)", 1, "error")
			} else if !known["citation_purpose"][purpose] {
				add("unknown_registry_value", path+".citation_purpose is not seeded; informational, not rejected (§12)", 8, "info")
			}
		}
		if rawCoordinates, present := ref["log_coordinates"]; present {
			coordinates := asMap(rawCoordinates)
			if coordinates == nil {
				add("reference_log_coordinates_malformed", path+".log_coordinates MUST be an object (§5.5.5)", 1, "error")
				continue
			}
			for _, field := range []string{"log_id", "leaf_index", "inclusion_proof"} {
				if value, ok := coordinates[field]; !ok || value == nil {
					add("reference_log_coordinates_malformed", path+".log_coordinates requires "+field+" (§5.5.5)", 1, "error")
				}
			}
		}
	}
	return findings
}

File: rust/bundle/src/references.rs
//! draft-04 §5.5.5 reference-block structural checks, without resolving cited
//! artifacts. Byte-for-byte port of the Go/Python/TS siblings' `references`
//! finding logic, minus the `citation_purpose` "known in registry" lookup:
//! that check is `info`-severity only (never gates `ok`) and depends on the
//! REGISTRY.md-derived vocabulary tables, which this crate does not embed
//! (see `verify.rs` module doc for the scope note). Every `error`-severity
//! reference finding is ported in full.

use crate::verify::{is_hex64, mkf, Finding};
use serde_json::{Map, Value};

pub fn reference_findings(capsule: &Map<String, Value>) -> Vec<Finding> {
    if !matches!(capsule.get("format_version"), Some(Value::String(v)) if v == "4") {
        return Vec::new();
    }
    let raw = match capsule.get("references") {
        Some(v) => v,
        None => return Vec::new(),
    };
    let mut findings = Vec::new();
    let refs = match raw.as_array() {
        Some(refs) => refs,
        None => {
            findings.push(mkf(
                "references_malformed",
                "references MUST be an array (§5.5.5)",
                Some(1),
                "error",
            ));
            return findings;
        }
    };
    let parent = capsule
        .get("chain")
        .and_then(Value::as_object)
        .and_then(|chain| chain.get("parent_capsule_id"))
        .and_then(Value::as_str);

    for (i, raw_ref) in refs.iter().enumerate() {
        let path = format!("references[{i}]");
        let reference = match raw_ref.as_object() {
            Some(m) => m,
            None => {
                findings.push(mkf(
                    "reference_malformed",
                    &format!("{path} MUST be an object (§5.5.5)"),
                    Some(1),
                    "error",
                ));
                continue;
            }
        };
        for field in ["type", "digest_alg", "digest"] {
            let ok = matches!(reference.get(field), Some(Value::String(s)) if !s.is_empty());
            if !ok {
                findings.push(mkf(
                    "reference_malformed",
                    &format!("{path}.{field} MUST be a non-empty string (§5.5.5)"),
                    Some(1),
                    "error",
                ));
            }
        }
        // Compare only a known AAC identity context, never equal-looking digests
        // belonging to a different artifact type or hash algorithm (CPB §7).
        let digest = reference
            .get("digest")
            .and_then(Value::as_str)
            .unwrap_or("");
        let is_aac_sha256 = matches!(reference.get("type"), Some(Value::String(t)) if t == "agent-action-capsule")
            && matches!(reference.get("digest_alg"), Some(Value::String(a)) if a == "SHA-256");
        if is_aac_sha256 {
            if !digest.is_empty() && !is_hex64(digest) {
                findings.push(mkf(
                    "reference_malformed",
                    &format!("{path}.digest MUST be an AAC Capsule ID for agent-action-capsule/SHA-256 (§5.5.5)"),
                    Some(1),
                    "error",
                ));
            }
            if let Some(parent) = parent {
                if !parent.is_empty() && digest == parent {
                    findings.push(mkf(
                        "reference_duplicates_chain_parent",
                        &format!("{path} duplicates chain.parent_capsule_id (§5.5.5)"),
                        Some(6),
                        "error",
                    ));
                }
            }
        }
        // citation_purpose registry-membership ("known"/"unknown") is info-only
        // and omitted here (see module doc); a present-but-empty/non-string
        // citation_purpose is still a structural error.
        if let Some(raw_purpose) = reference.get("citation_purpose") {
            let ok = matches!(raw_purpose, Value::String(s) if !s.is_empty());
            if !ok {
                findings.push(mkf(
                    "reference_malformed",
                    &format!("{path}.citation_purpose MUST be a non-empty string (§5.5.5)"),
                    Some(1),
                    "error",
                ));
            }
        }
        if let Some(raw_coordinates) = reference.get("log_coordinates") {
            let coordinates = match raw_coordinates.as_object() {
                Some(m) => m,
                None => {
                    findings.push(mkf(
                        "reference_log_coordinates_malformed",
                        &format!("{path}.log_coordinates MUST be an object (§5.5.5)"),
                        Some(1),
                        "error",
                    ));
                    continue;
                }
            };
            for field in ["log_id", "leaf_index", "inclusion_proof"] {
                if !matches!(coordinates.get(field), Some(v) if !v.is_null()) {
                    findings.push(mkf(
                        "reference_log_coordinates_malformed",
                        &format!("{path}.log_coordinates requires {field} (§5.5.5)"),
                        Some(1),
                        "error",
                    ));
                }
            }
        }
    }
    findings
}
